core/node/node_framework/src/service/named_future.rs

```diff
@@ -0,0 +1,52 @@
+use std::{fmt, future::Future, pin::Pin, task};
+
+use pin_project_lite::pin_project;
+
+use crate::task::TaskId;
+
+pin_project! {
+    /// Implements a future with the name tag attached.
+    pub struct NamedFuture<F> {
+        #[pin]
+        inner: F,
+        name: TaskId,
+    }
+}
+
+impl<F> NamedFuture<F>
+where
+    F: Future,
+{
+    /// Creates a new future with the name tag attached.
+    pub fn new(inner: F, name: TaskId) -> Self {
+        Self { inner, name }
+    }
+
+    pub fn id(&self) -> TaskId {
+        self.name.clone()
+    }
+
+    pub fn into_inner(self) -> F {
+        self.inner
+    }
```

Comment:
Bikeshedding: Since this is only used to spawn hooks onto a Tokio runtime, it could make sense to define a dedicated method instead:

```rust
impl<F> NamedFuture<F>
where
    F: Future + Send + 'static,
    F::Output: Send + 'static,
{
    pub fn spawn(self, handle: &tokio::runtime::Handle) -> NamedFuture<JoinHandle<F::Output>> {
        NamedFuture {
            name: self.name,
            inner: handle.spawn(self.inner),
        }
    }
}
```

Reply:
That makes sense, will do it in the next PR!